model_compression_toolkit/core/pytorch/hessian/weights_trace_hessian_calculator_pytorch.py

```diff
@@ -14,37 +14,109 @@
+                                                                   num_iterations_for_approximation=num_iterations_for_approximation)
+
+
+    def compute(self) -> np.ndarray:
+        """
+        Compute the Hessian-based scores w.r.t target node's weights.
+
+        Returns:
+            The computed scores as numpy ndarray for target node's weights.
+        """
+
+        # Check if the target node's layer type is supported
+        if not DEFAULT_PYTORCH_INFO.is_kernel_op(self.hessian_request.target_node.type):
+            Logger.error(f"{self.hessian_request.target_node.type} is not supported for Hessian info w.r.t weights.")  # pragma: no cover
+
+        # Float model
+        model, _ = FloatPyTorchModelBuilder(graph=self.graph).build_model()
+
+        # Get the weight attributes for the target node type
+        weights_attributes = DEFAULT_PYTORCH_INFO.get_kernel_op_attributes(self.hessian_request.target_node.type)
+
+        # Get the weight tensor for the target node
+        if len(weights_attributes) != 1:
+            Logger.error(f"Hessian scores w.r.t weights is supported, for now, for a single-weight node. Found {len(weights_attributes)}")
+
+        weights_tensor = getattr(getattr(model,self.hessian_request.target_node.name),weights_attributes[0])
+
+        # Get the output channel index
+        output_channel_axis, _ = DEFAULT_PYTORCH_INFO.kernel_channels_mapping.get(self.hessian_request.target_node.type)
+        shape_channel_axis = [i for i in range(len(weights_tensor.shape))]
+        if self.hessian_request.granularity == HessianInfoGranularity.PER_OUTPUT_CHANNEL:
+            shape_channel_axis.remove(output_channel_axis)
+        elif self.hessian_request.granularity == HessianInfoGranularity.PER_ELEMENT:
+            shape_channel_axis = ()
+
+        # Run model inference
+        outputs = model(self.input_images)
+        output_tensor = self.concat_tensors(outputs)
+        device = output_tensor.device
+
+        approximation_per_iteration = []
+        for j in range(self.num_iterations_for_approximation):
+            # Getting a random vector with normal distribution and the same shape as the model output
+            v = torch.randn_like(output_tensor, device=device)
+            f_v = torch.mean(torch.sum(v * output_tensor, dim=-1))
+            # Compute gradients of f_v with respect to the weights
+            f_v_grad = autograd.grad(outputs=f_v,
+                                     inputs=weights_tensor,
+                                     retain_graph=True)[0]
+
+            # Trace{A^T * A} = sum of all squares values of A
+            approx = f_v_grad ** 2
+            if len(shape_channel_axis) > 0:
+                approx = torch.sum(approx, dim=shape_channel_axis)
+
+            if j > MIN_JACOBIANS_ITER:
+                new_mean = (torch.sum(torch.stack(approximation_per_iteration))+approx)/(j+1)
```

Comment:
Are you sure it computes the mean in the per-channel and per-element case? If it's I would add the axis to the sum just to make it clearer

Reply:
you are right. I forgot the axis.